### ISSUE TITLE
`self` and `static` return types don't work if parent class has a constrained template type

### ISSUE THREAD

**Issue by janopae** [vimeo/psalm#10476]
Consider the following code: https://psalm.dev/r/c676544a45

It works as expected: The parent class `get` method returns `static` – therefore, when called on the child, we can be sure that the `self` return type is correct.

Now, let's constrain the template parameter on the parent class (but in a way where the child is still valid). Urgh! https://psalm.dev/r/6f35cf8914

Now Psalm complains that "`The type 'ParentClass<object>' is more general than the declared return type 'ChildClass' for ChildClass::myGet`".

We can't fix that by using `static` on the child return type: https://psalm.dev/r/091f9485a7

Seems like we can't fix it at all in our application code, since Psalm seems to override the type constraint over the actual known type of the template.

**Comment by weirdan** [comment]
The fix seems to be rather simple: https://psalm.dev/r/872aced924

**Comment by janopae** [comment]
Thanks, @weirdan! I didn't think of a solution like this, because in my case, the parent class lived in a library.

Do you think Psalm should change its behaviour in this regard? After all, the resolved `static` type doesn't even support template types, so `static<T>` doesn't make so much sense. Otherwise, feel free to close this issue.

**Comment by tamas-marton** [comment]
I agree with @janopae. While the solution of @weirdan fixes this, it's:

- not possible (or complicated) when using a library
- not the expected behaviour from Psalm

If it's difficult to sovle in Psalm, it should be mentioned in the docs, on the [templating page](https://psalm.dev/docs/annotating_code/templated_annotations/#template-constraints).

PS: Another example demonstrating my case: https://psalm.dev/r/facad6320f
